restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/OrganizationsEndpoint.java

```diff
@@ -0,0 +1,253 @@
+        checkAuthenticatedAccount();
+        allowOnlySuperAdmin();
+
+        List<Organization> organizations = null;
+
+        String status = info.getQueryParameters().getFirst("Status");
+
+        if(status != null && Organization.Status.getValueOf(status.toLowerCase()) != null){
+            organizations = organizationsDao.getOrganizationsByStatus(Organization.Status.getValueOf(status.toLowerCase()));
+        }else{
+            organizations = organizationsDao.getAllOrganizations();
+        }
+        if (organizations == null || organizations.isEmpty()) {
+            return status(NOT_FOUND).build();
+        }
+        if (APPLICATION_XML_TYPE == responseType) {
+            final RestCommResponse response = new RestCommResponse(new OrganizationList(organizations));
+            return ok(xstream.toXML(response), APPLICATION_XML).build();
+        } else if (APPLICATION_JSON_TYPE == responseType) {
+            return ok(gson.toJson(organizations), APPLICATION_JSON).build();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * putOrganization create new organization
+     * @param domainName
+     * @param data
+     * @param applicationJsonType
+     * @return
+     */
+    protected Response putOrganization(String domainName, MultivaluedMap<String, String> data,
+            MediaType responseType) {
+        if(domainName == null){
+            return status(BAD_REQUEST).entity(MSG_EMPTY_DOMAIN_NAME ).build();
+        }else{
+            checkAuthenticatedAccount();
+            allowOnlySuperAdmin();
+
+            //Character verification
+            final Pattern pattern = Pattern.compile(SUB_DOMAIN_NAME_VALIDATION_PATTERN);
+            if(!pattern.matcher(domainName).matches()){
+                return status(BAD_REQUEST).entity(MSG_INVALID_DOMAIN_NAME_PATTERN).build();
+            }
+
+            //Check if domain_name does not already taken inside restcomm by an organization.
+            Organization organization = organizationsDao.getOrganizationByDomainName(domainName);
+            if(organization != null){
+                return status(CONFLICT)
+                        .entity(MSG_DOMAIN_NAME_NOT_AVAILABLE)
+                        .build();
+            }
+            if(dnsProvisioningManager == null) {
+                logger.warn("No DNS provisioning Manager is configured, restcomm will not make any queries to DNS server.");
+                organization = new Organization(Sid.generate(Sid.Type.ORGANIZATION), domainName, DateTime.now(), DateTime.now(), Organization.Status.ACTIVE);
+                organizationsDao.addOrganization(organization);
+            }else {
+                //for example hosted zone id of domain restcomm.com or others. if not provided then default will be used as per configuration
+                String hostedZoneId = data.getFirst("HostedZoneId");
+                if(dnsProvisioningManager.doesResourceRecordAlreadyExists(domainName, hostedZoneId)){
```

Comment:
again, is there any kind of backend transaction happening here? i didnt see any transaction at dao impl level, is this at amazon backend level? shouldnt we synchronize the conurrent acces to ensure this kind of unique constraints are safe?

Reply:
This particular line is making sure if new domain name is available on dns/amazon-roue53 side